### PR TITLE
Scala 3: Add support to format open classes.

### PR DESCRIPTION
**Pull request by olafurpg**
This PR implements support to format open classes, which is a new soft
modifier in Scala 3. Details
https://dotty.epfl.ch/docs/reference/other-new-features/open-classes.html

### CODE REVIEW

**Review by tgodzik — APPROVED** [review]
LGTM! And we thought it would be a super quick feature :sweat_smile:

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala [hunk 1]
```diff
@@ -87,10 +87,13 @@ object ScalafmtRunner {
     val scala213 = Scala213
       .withAllowTrailingCommas(true)
     val default = scala213
-      .withAllowOrTypes(true) // New feature in Dotty
-      .withAllowAndTypes(true) // New feature in Dotty
-      .withAllowTraitParameters(true) // New feature in Dotty
-      .withAllowColonForExtractorVarargs(true) // New feature in Dotty
+      // Supported Dotty features.
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala:90]: unrelated question: for the other changes that @MichalBednarz and @tgodzik have been working on (like `using` and `enum`), do we need to add these properties?

I can see one for `.withAllowEnums`, and something about `.withAllowGivenUsing` (although i am not sure if that's the same as `using`).

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala [hunk 2]
```diff
@@ -1,24 +1,34 @@
...
-    )
+  final case class ModKey(
+      name: String,
+      matches: Mod => Boolean
+  ) extends Product
+  private def createMod[T <: Mod](implicit
+      name: Name,
+      c: Classifier[Tree, T]
+  ): ModKey = ModKey(name.value, m => c(m))
+  val `private` = createMod[Mod.Private]
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:23]: i'd add an empty line before this

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala [hunk 3]
```diff
@@ -1,24 +1,34 @@
...
-      `private`,
-      `protected`,
-      `lazy`
-    )
+  final case class ModKey(
+      name: String,
+      matches: Mod => Boolean
+  ) extends Product
+  private def createMod[T <: Mod](implicit
+      name: Name,
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:20]: also, question (on scala, probably): how does this work? why is there a unique implicit `Name` and `Classifier`?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala [hunk 4]
```diff
@@ -1,24 +1,34 @@
...
-      `abstract`,
-      `override`,
-      `private`,
-      `protected`,
-      `lazy`
-    )
+  final case class ModKey(
+      name: String,
+      matches: Mod => Boolean
+  ) extends Product
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:18]: a case class always extends `Product`, no?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala [hunk 5]
```diff
@@ -1,24 +1,34 @@
...
+  private def createMod[T <: Mod](implicit
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:19]: i'd add an empty line before this

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala [hunk 6]
```diff
@@ -1,24 +1,34 @@
...
+      name: Name,
+      c: Classifier[Tree, T]
+  ): ModKey = ModKey(name.value, m => c(m))
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:22]: why not `c.apply`?

**Review by kitbellew — CHANGES_REQUESTED** [review]
there's some scala magic happening here :) but a few questions still...

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala (hunk 3 above)
Comment (poslegm) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:20]: there is [macro machinery](https://github.com/lihaoyi/sourcecode/blob/master/sourcecode/src-2/sourcecode/Macros.scala#L72-L77) around it

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala (hunk 1 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala:90]: I think we will use the separate Dotty dialect for now (until parser is fully stable) and require users to add `runner.dialect`. I think @olafurpg already included it in another PR for macros.

**Comment by olafurpg** [comment]
Superseded by #2286, where this commit https://github.com/scalameta/scalafmt/pull/2286/commits/6e0878fb8654e3cb64a97655e91bed3c032cfb0c addresses the comments from this PR

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala (hunk 4 above)
Comment (olafurpg) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:18]: Correct, removed.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala (hunk 5 above)
Comment (olafurpg) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:19]: Removed this method. Inserted a blank line after the class definition.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala (hunk 6 above)
Comment (olafurpg) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/SortSettings.scala:22]: No specific reason. I admit this method was unnecessarily complicated so I removed the method.

**Pull request closed without merging**